stream/distributedlog/core/src/main/java/org/apache/distributedlog/ReadAheadEntryReader.java

```diff
@@ -942,25 +948,60 @@ private void unsafeMoveToNextLogSegment() {
                         readHandler.getFullyQualifiedName());
                 isCatchingUp = false;
             }
-            pauseReadAheadOnNoMoreLogSegments();
         }
+
+        scheduleRead();
     }
 
-    void scheduleReadNext() {
+    void completeRead() {
         orderedSubmit(new CloseableRunnable() {
-            @Override
-            public void safeRun() {
-                if (null == currentSegmentReader) {
-                    pauseReadAheadOnNoMoreLogSegments();
-                    return;
+                @Override
+                public void safeRun() {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("[{}][state:{}] Read completed", streamName, state);
+                    }
+                    if (state == State.READING) {
+                        state = State.IDLE;
+                    }
                 }
-                unsafeReadNext(currentSegmentReader);
-            }
-        });
+            });
     }
 
-    private void unsafeReadNext(SegmentReader reader) {
-        reader.readNext().whenComplete(this);
+    void scheduleRead() {
+        orderedSubmit(new CloseableRunnable() {
+                @Override
+                public void safeRun() {
+
+                    boolean cacheFull = isCacheFull();
+                    SegmentReader reader = currentSegmentReader;
+                    boolean hasMoreSegments = reader != null;
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("[{}][state:{}] scheduling read, cacheFull {}, hasMoreSegments {}",
+                                     streamName, state, cacheFull, hasMoreSegments);
+                    }
+                    switch (state) {
+                    case IDLE:
+                        if (cacheFull || !hasMoreSegments) {
+                            state = State.PAUSED;
+                        } else {
+                            reader.readNext().whenComplete(ReadAheadEntryReader.this);
+                            state = State.READING;
+                        }
+                        break;
+                    case PAUSED:
+                        if (!cacheFull && hasMoreSegments) {
+                            reader.readNext().whenComplete(ReadAheadEntryReader.this);
+                            state = State.READING;
+                        }
+                        break;
+                    case READING:
+                    case ERROR:
+                    case CLOSED:
+                        // do nothing
```

Comment:
shall a log at trace level be useful ?

Reply:
nah, we already know if we're going to hit this branch from the debug log above.